aws/credentials/process_provider.go

```diff
@@ -0,0 +1,228 @@
+package credentials
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/internal/shareddefaults"
+	"github.com/go-ini/ini"
+)
+
+// ProcessProviderName provides a name of Process provider
+const ProcessProviderName = "ProcessProvider"
+
+var (
+	// ErrProcessProviderHomeNotFound is emitted when the user directory cannot be found.
+	ErrProcessProviderHomeNotFound = awserr.New("UserHomeNotFound", "user home directory not found.", nil)
+)
+
+// A ProcessProvider retrieves credentials from a process specified as
+// 'credential_process' in a profile, and keeps track if those credentials are
+// expired.
+//
+// Profile ini file example: $HOME/.aws/config
+type ProcessProvider struct {
+	Expiry
+
+	// Path to the shared config file.
+	//
+	// If empty will look for "AWS_CONFIG_FILE" env variable. If the
+	// env value is empty will default to current user's home directory.
+	// Linux/OSX: "$HOME/.aws/config"
+	// Windows:   "%USERPROFILE%\.aws\config"
+	Filename string
+
+	// AWS Profile to retrieve the 'credential_process' from. If empty
+	// will default to environment variable "AWS_PROFILE" or "default" if
+	// environment variable is also not set.
+	Profile string
+
+	// ExpiryWindow will allow the credentials to trigger refreshing prior to
+	// the credentials actually expiring. This is beneficial so race conditions
+	// with expiring credentials do not cause request to fail unexpectedly
+	// due to ExpiredTokenException exceptions.
+	//
+	// So a ExpiryWindow of 10s would cause calls to IsExpired() to return true
+	// 10 seconds before the credentials are actually expired.
+	//
+	// If ExpiryWindow is 0 or less it will be ignored.
+	ExpiryWindow time.Duration
+
+	// A function to perform the credential_process execution. Can be replaced
+	// for testing
+	executionFunc func(process string) ([]byte, error)
+}
+
+// NewProcessProvider returns a pointer to a new Credentials object
+// wrapping the Profile file provider.
+func NewProcessProvider(filename, profile string) *Credentials {
+	return NewCredentials(&ProcessProvider{
+		Filename:      filename,
+		Profile:       profile,
+		executionFunc: executeCredentialProcess,
+	})
+}
+
+// Retrieve executes the 'credential_process' and returns the credentials
+func (p *ProcessProvider) Retrieve() (Value, error) {
+	filename, err := p.filename()
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, err
+	}
+
+	creds, err := p.loadProfile(filename, p.profile())
```

Comment:
Loading the shared config file each time the credentials are retrieved have issues that I don't think is good for the SDK to perform. per #1993